bsp/worker/src/mill/bsp/worker/Utils.scala

```diff
@@ -46,6 +58,75 @@ private object Utils {
     else StatusCode.OK
   }
 
+  def makeBuildTarget(
+      id: BuildTargetIdentifier,
+      depsIds: Seq[BuildTargetIdentifier],
+      bt: BspBuildTarget,
+      data: Option[(String, Object)]
+  ): BuildTarget = {
+    val buildTarget = new BuildTarget(
+      id,
+      bt.tags.asJava,
+      bt.languageIds.asJava,
+      depsIds.asJava,
+      new BuildTargetCapabilities().tap { it =>
+        it.setCanCompile(bt.canCompile)
+        it.setCanTest(bt.canTest)
+        it.setCanRun(bt.canRun)
+        it.setCanDebug(bt.canDebug)
+      }
+    )
+
+    bt.displayName.foreach(buildTarget.setDisplayName)
+    bt.baseDirectory.foreach(p => buildTarget.setBaseDirectory(sanitizeUri(p)))
+
+    for ((dataKind, data) <- data) {
+      buildTarget.setDataKind(dataKind)
+      buildTarget.setData(data)
+    }
+    buildTarget
+  }
+
+  def outputPaths(
+      outDir: os.Path,
+      buildTargetBaseDir: os.Path,
+      topLevelProjectRoot: os.Path
+  ): Seq[OutputPathItem] = {
+    val output = new OutputPathItem(
+      // Spec says, a directory must end with a forward slash
+      sanitizeUri(outDir) + "/",
+      OutputPathItemKind.DIRECTORY
+    )
+
+    def ignore(path: os.Path): Boolean = {
+      path.last.startsWith(".") ||
+      path.endsWith(os.RelPath("out")) ||
+      path.endsWith(os.RelPath("target")) ||
+      path.endsWith(os.RelPath("docs")) ||
+      path.endsWith(os.RelPath("mill-build"))
+    }
+
+    def projectsRootPaths = os.walk(topLevelProjectRoot, ignore).collect {
+      case p if p.endsWith(os.RelPath("build.sc")) => p / os.up
+    }
+    def outputPathItem(path: os.Path) =
+      new OutputPathItem(
+        // Spec says, a directory must end with a forward slash
+        sanitizeUri(path) + "/",
+        OutputPathItemKind.DIRECTORY
+      )
+    def additionalExclusions = projectsRootPaths.flatMap { path =>
```

Comment:
Got it. Is everything in `buildTargetOutputPaths` considered excluded?

Reply:
Formally, according to BSP docs it's BSP client decision what to do with buildOutputPaths response from BSP server. most of clients are excluding them, in particular intellij. 
What's more ,in case of intellij there are the limitations I mentioned in the PR description (these are not formalized though ,it's just what I have observed by reading through intellij bsp source code and confirmed by testing )

Reply:
Got it. Can you also manually test VSCode+Metals to see what the behavior is there? Just so we know what affect this PR will have on those users (if any)